src/gt4py/stencil_object.py

```diff
@@ -1,27 +1,47 @@
 # -*- coding: utf-8 -*-
+#
+# GT4Py - GridTools4Py - GridTools for Python
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
 import abc
 import collections.abc
 import sys
 import time
 import warnings
-from typing import Any, Collection, Dict, Tuple
+from typing import Any, Callable, Dict, Optional, Tuple, Union
 
 import numpy as np
 
 import gt4py.backend as gt_backend
-import gt4py.ir as gt_ir
 import gt4py.storage as gt_storage
 import gt4py.utils as gt_utils
-from gt4py.definitions import (
-    AccessKind,
-    Boundary,
-    CartesianSpace,
-    DomainInfo,
-    FieldInfo,
-    Index,
-    ParameterInfo,
-    Shape,
-)
+from gt4py.definitions import DomainInfo, FieldInfo, Index, ParameterInfo, Shape
+
+
+FieldType = Union[gt_storage.storage.Storage, np.ndarray]
+OriginType = Union[Tuple[int, int, int], Dict[str, Tuple[int, ...]]]
+
+
+def _compute_cache_key(field_args, parameter_args, domain, origin) -> int:
+    field_hashes = tuple(
+        hash(name) + hash(field.shape) + hash(field.default_origin)
+        for name, field in field_args.items()
+    )
+    parameter_hashes = tuple(hash(name) + hash(param) for name, param in parameter_args.items())
+    domain_hash = hash(domain)
+    origin_hash = hash(origin)
+    return hash((*field_hashes, *parameter_hashes, domain_hash, origin_hash))
```

Comment:
`hash()` should only be called **_once_** with an argument tuple containing all the information needed to create the key.